sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -732,34 +750,88 @@ object GpuCast extends Arm {
           ColumnVector.stringConcatenate(empty, nullRep, Array(leftColumn, strVec, rightColumn))
         }
       }
-      /* -------------------------------- helper functions -----------------------*/
 
+      val strChildContainsNull = withResource(input.getChildColumnView(0)) {child =>
+        doCast(
+          child, elementType, StringType, ansiMode, legacyCastToString, stringToDateAnsiModeEnabled)
+      }
 
-      // cast child column to string type
-      withResource(input.getChildColumnView(0)) { childView =>
-        withResource(castChildToStr(childView)){ stringChild =>
-          withResource(input.replaceListChild(stringChild)) {strArr =>
-            // concatenate each row. cast from list column to string column
-            withResource(ColumnVector.fromScalar(sep, numRows)){ sepCol =>
-              withResource(
-                strArr.stringConcatenateListElements(sepCol)
-              ) { strColContainsNull =>
-                withResource(strColContainsNull.replaceNulls(empty)){strCol =>
-                  withResource(removeFirstSpace(strCol)){withoutBrackets =>
-                    withResource(addBrackets(withoutBrackets))(
-                      _.mergeAndSetValidity(BinaryOp.BITWISE_AND, input)
-                    )
-                  }
-                }
-              }
+      withResource(strChildContainsNull) {strChildContainsNull =>
+        withResource(input.replaceListChild(strChildContainsNull)) {strArrayCol =>
+          withResource(concatenateStringArrayElements(strArrayCol, legacyCastToString)) {strCol =>
+            withResource(addBrackets(strCol)) {
+              _.mergeAndSetValidity(BinaryOp.BITWISE_AND, input)
             }
           }
         }
       }
     }
   }
 
+  private def castMapToString(
+      input: ColumnView,
+      from: MapType,
+      ansiMode: Boolean,
+      legacyCastToString: Boolean,
+      stringToDateAnsiModeEnabled: Boolean): ColumnVector = {
 
+    val numRows = input.getRowCount.toInt
+    val (arrowStr, emptyStr, spaceStr) = ("->", "", " ")
+    val (leftStr, rightStr, nullStr) =
+      if (legacyCastToString) ("[", "]", "") else ("{", "}", "null")
+
+    val (strKey, strValue) = withResource(input.getChildColumnView(0)) { kvStructColumn =>
+      val strKey = withResource(kvStructColumn.getChildColumnView(0)) { keyColumn =>
+        doCast(
+          keyColumn, from.keyType, StringType, ansiMode,
+          legacyCastToString, stringToDateAnsiModeEnabled)
+      }
+      val strValue = withResource(kvStructColumn.getChildColumnView(1)) { valueColumn =>
```

Comment:
nit: Technically if an exception happens when computing `strValue` `strKey` will be leaked. Can we have a `closeOnExcept` for strKey while computing `strValue`?

Reply:
Done